Linux 2025

danoli3 opened the pull request:
## Linux
- Updates for GCC14 / GCC13 / GCC12 / GCC11 / GCC10 (from GCC6 std=c++11) 
- C++ now variant to C++23 depending GCC version. 24.04 = GCC14
- Download Script now updated to download correct GCC based on whats installed. 
- Performance Boost for Linux apps 400% for correct GCC 
- Split Linux64 to -TARGET = Linux - A = 64
- Split allows for all versions of Linux to fall into correct locations (arm64 etc) 
- Updated Workflow for updated Linux Libraries 
- Fixed Nightly borked

danoli3 committed:
Submodule fixes.

danoli3 committed:
Updates nightly

danoli3 committed:
Download libs [ 2.6.4 ~ 2.7.0]  - Added GCC / Linux updates

danoli3 committed:
remove gold linux binutils ?

danoli3 committed:
Linux Workflow updates / mkfile

danoli3 commented:
Removed Gold LD for the moment regarding:
https://github.com/openframeworks/openFrameworks/issues/8239

danoli3 committed:
Script moving oh no

danoli3 committed:
Linux Script Updates

danoli3 committed:
Gold LD if found

danoli3 committed:
Downloader [2.7.1 ~ 2.7.2] exit error if fail on download of library rather than let continue to code compiltion errors
emscripten memory64 renamed 64 as internals are em32 and em64

danoli3 committed:
Downloader [2.7.2 ~ 2.7.3]

danoli3 committed:
Linux Actions

danoli3 committed:
download_libs [2.7.3 ~ 2.7.4] fix opt args for downloads Linux

danoli3 committed:
downloader set linux aarch64 to 64 for testing on arm64 linux

danoli3 committed:
Makefile Linux subpath

danoli3 committed:
makefile subpath

danoli3 committed:
Linux subpath

danoli3 committed:
download libs validate downloads before overrwite and 2.8.0. wget2 disabled for the moment

danoli3 committed:
Makefiles

danoli3 committed:
downloadlibs [2.8.0 ~ 2.8.1] valid bool to int

danoli3 committed:
ci linux

danoli3 committed:
Workflow ci optimisation VS single ARCH downloads
Actions added 22.04 linux GCC10 test

danoli3 committed:
Linux 22.04 Test

danoli3 committed:
VS download optimisation

danoli3 committed:
VS Download fix

danoli3 committed:
Makefile fixes / vs arch

danoli3 commented:
Makefile linux:
```
fi
building with makefiles
**** Building OF core ****
  HOST_OS=Linux
  PLATFORM_ARCH=x86_64
  PLATFORM_OS=Linux
  HOST_ARCH=x86_64
  HOST_OS=Linux
  CROSS_COMPILING=0
  PLATFORM_VARIANT=default
  IS_RASPBIAN=
  =================== config.mk platform detection ================
  PLATFORM_ARCH=x86_64
  PLATFORM_OS=Linux
  PLATFORM_VARIANT=default
  PLATFORM_LIB_SUBPATH=linux/64
  =================== config.mk paths =============================
  OF_ADDONS_PATH=../../../addons
  OF_EXAMPLES_PATH=../../../examples
  OF_APPS_PATH=../../../apps
  OF_LIBS_PATH=../../../libs
  OF_LIBS_OPENFRAMEWORKS_PATH=../../../libs/openFrameworks
  OF_LIBS_OF_COMPILED_PATH=../../../libs/openFrameworksCompiled
  OF_LIBS_OF_COMPILED_PROJECT_PATH=../../../libs/openFrameworksCompiled/project
  OF_SHARED_MAKEFILES_PATH=./makefileCommon
  OF_PLATFORM_MAKEFILES=../../../libs/openFrameworksCompiled/project/linux/64
  OF_CORE_LIB_PATH=../../../libs/openFrameworksCompiled/lib/linux/64
  makefileCommon/config.shared.mk:259: *** recipe commences before first target.  Stop.
  Error: Process completed with exit code 2.
  ```
  
  Really need to get the cmake stuff in soon. 
  
  I'd re-added gold ld again with command check.
   - 
   - todo Add to install libs setup

danoli3 commented:
Latest libraries now has Linux arm64:
https://github.com/openframeworks/apothecary/releases/tag/latest

Linux 64

danoli3 committed:
Makefile depth fix

danoli3 committed:
Makefile strict tab

danoli3 committed:
Github Actions Test needs

danoli3 committed:
Addons linux path ofxSvg make

danoli3 committed:
Actions needs use ubuntu as quick

danoli3 committed:
Actions rename Linux-64 to linux

danoli3 committed:
Makefiles spaces to tabs

danoli3 committed:
Makefile debug

danoli3 committed:
Reset file

danoli3 committed:
Spaces to Tabs

danoli3 committed:
no debug flags

danoli3 committed:
Actions

danoli3 commented:
```
Getting action download info
Error: This request has been automatically failed because it uses a deprecated version of `actions/upload-artifact: v3`. Learn more: https://github.blog/changelog/2024-04-16-deprecation-notice-v3-of-the-artifact-actions/
```
![image](https://github.com/user-attachments/assets/dbc48b49-8ff7-4a97-b84a-ef2d087a304a)

Weird. Disabling caches / updating checkout to 4.2.2 
Added workflow depend on linux for others (its fastest and saves cpu - once its working)

danoli3 committed:
22.04 disabled

danoli3 commented:
Linux empty example is compiling just not addonsExample

danoli3 committed:
ci build 64

danoli3 committed:
Line spaces to tabs

danoli3 committed:
gold

ofTheo commented:
Looks like PLATFORM_CORE_EXLUSIONS is not defined in config.shared.mk

![image](https://github.com/user-attachments/assets/c94359c7-9795-4d54-8bde-75daba3d6963)

danoli3 committed:
Fix Linux string issue

danoli3 committed:
debug off

danoli3 committed:
gold linker on command

danoli3 commented:
@ofTheo yeah the library subpath was being used for the targetarch name in for the makefile in the directory! So with linux/64 it was bugging out. I thought it was line spacing tab stuff.

Makefiles are so sensitive though wow. It looks like maybe the streamer issue just reported is a spaces instead of tabs issue if they had run the code. 

```
  ccache g++ -o bin/linux64_debug  obj/linux/64/Debug/src/ofApp.o obj/linux/64/Debug/src/main.o   ../../../libs/openFrameworksCompiled/lib/linux/64/libopenFrameworksDebug.a  -Wl,-rpath=./libs:./bin/libs:./ -Wl,--as-needed -Wl,--gc-sections -fuse-ld=gold ../../../libs/brotli/lib/linux/64/libbrotlienc.a ../../../libs/brotli/lib/linux/64/libbrotlicommon.a ../../../libs/brotli/lib/linux/64/libbrotlidec.a ../../../libs/fmt/lib/linux/64/libfmt.a ../../../libs/glew/lib/linux/64/libGLEW.a ../../../libs/glfw/lib/linux/64/libglfw3.a ../../../libs/tess2/lib/linux/64/libtess2.a ../../../libs/zlib/lib/linux/64/zlib.a -lgstapp-1.0 -lgstvideo-1.0 -lgstbase-1.0 -lgstreamer-1.0 -ludev -lfontconfig -lfreetype -lsndfile -lopenal -lssl -lcrypto -lcurl -lrtaudio -lpulse-simple -lpulse -pthread -lasound -L/usr/lib64 -lGLEW -lGL -lX11 -lGLU -lgtk-3 -lgdk-3 -lz -lpangocairo-1.0 -lpango-1.0 -lharfbuzz -latk-1.0 -lcairo-gobject -lcairo -lgdk_pixbuf-2.0 -lgio-2.0 -lgobject-2.0 -lglib-2.0 -lmpg123  -lglut -lX11 -lXrandr -lXxf86vm -lXi -lXcursor -ldl -lpthread -lfreeimage -lpugixml -luriparser
  ../../../libs/openFrameworks/sound/ofOpenALSoundPlayer.cpp:1009: error: undefined reference to 'kiss_fftr_alloc'
  ../../../libs/openFrameworks/sound/ofOpenALSoundPlayer.cpp:1020: error: undefined reference to 'kiss_fftr_alloc'
  ../../../libs/openFrameworks/sound/ofOpenALSoundPlayer.cpp:1065: error: undefined reference to 'kiss_fftr'
  ../../../libs/openFrameworks/sound/ofOpenALSoundPlayer.cpp:1095: error: undefined reference to 'kiss_fftr'
  collect2: error: ld returned 1 exit status
  ```
  
  Linking Kiss is missing! Sweet looking into

danoli3 committed:
gold linker comments disabled

danoli3 committed:
Update addon_config

danoli3 committed:
Actions Linux 22.04 Enabled

danoli3 committed:
Actions Linux 22.04 GCC 10

danoli3 commented:
# Complete 

  Ubuntu
  24.04.1
  LT
  Build Success all GCC targets
  
  Ubuntu
  22.04.1
  LT
  Build Success GCC10 (max GCC for 22.04)

danoli3 commented:
Added a utility pipe to apothecary and just did the same for oF
<img width="572" alt="image" src="https://github.com/user-attachments/assets/c4010d7e-96df-46e4-92a9-34fa593d15bb" />

this will make everything a lot easier to update before the gui stuff is done 
./of or ./of.sh 
usage following our function names: setup, update
```of setup``` runs any setup scripts if applicable, apt gets etc
```of update``` runs any download libs scripts

currently just targeting auto the platform that it is. 
macOS will call default macos
Windows will call the default scripts for vs 

This will drop clicks / folder clicks and bash running into 1 open repo in terminal: boom run command

danoli3 committed:
downloader [4.4.0 ~ 4.5.0] - fix close connection for zero urls.
shfmt formatted bash

danoli3 committed:
Actions update
Rpi disabled while fixed

ofTheo commented on addons/ofxAssimpModelLoader/addon_config.mk:
```diff
@@ -80,7 +80,7 @@ linux:
 	ADDON_INCLUDES_EXCLUDE = libs/assimp/%
 
-linux64:
+linux/64:
```
Will the old `linux64` and `linuxarmv6l` still work for PG etc?

ofTheo commented on addons/ofxSvg/addon_config.mk:
```diff
@@ -68,21 +68,21 @@ ios:
 
-linux64:
-	ADDON_LIBS = libs/svgtiny/lib/linux64/libsvgtiny.a
-	ADDON_LIBS += libs/libxml2/lib/linux64/libxml2.a
```
Likewise will libs still get picked up in older addons from `lib/linux64`

ofTheo commented on libs/openFrameworksCompiled/project/makefileCommon/config.shared.mk:
```diff
@@ -70,32 +70,39 @@ endif
 	ifeq ($(PLATFORM_OS),Linux)
 		ifeq ($(PLATFORM_ARCH),x86_64)
-			PLATFORM_LIB_SUBPATH=linux64
+			PLATFORM_LIB_SUBPATH=linux/64
```
looks like older projects with addons in lib/linux64/ might not work based on this?

danoli3 committed:
Linux script updates

danoli3 commented:
Good idea although make file location is a very minimal issue compared to all the ```std::``` upgrades required for older projects... but I've added the beginning on an auto-upgrade script.

this script will automatically replace issue strings in ```addon_config.mk``` for Linux.

Called via:
```./of upgrade addons```
```./of upgrade apps```


In the future this can also add in std:: replacements. Just requires stop and prompt by user since they need to back up files.
Backup mk files are created and removed only upon user input

ofTheo commented:
Upgrade script are always handy, but I think we have to be able to support addons that would be fully compliant up until this PR is merged. 

I know its a pain to support two names, but I'd either suggest leaving the current system of linux64, linuxarmv6l or, making sure both linux/64 and linux64 libs get picked up by make files / PG etc. 

This feels like the amount it will break, is not worth the conceptual tidiness of having the subfolder approach. 

Curious what others think, but I am def nervous as it's the osx / macOS issue again, but with even more platforms.

danoli3 commented:
Actually all those issues were caused by explicit makefile definitions like this one, when we coded not to at all use them! thanks for reminder ~ these are redundant now. Even the exclude code should be removed and just use actual platform defines. 

In fact lets just remove these completely as the projectGenerator figures this out automatically.

Here's a test fully removing these .a explicit:
<img width="1379" alt="image" src="https://github.com/user-attachments/assets/b712b17f-367f-4a1c-b43e-e5bc8296e92b" />

 Xcode:
<img width="931" alt="image" src="https://github.com/user-attachments/assets/4546f1e0-3fad-48e2-9e6b-2d3eb051f47d" />
Make:
<img width="578" alt="image" src="https://github.com/user-attachments/assets/26c28b0f-aef2-4e3e-8bf4-8571c98cbfd8" />

Let me test this on native linux see if it just works without this

danoli3 committed:
osx / ios remove defined .a

danoli3 committed:
osx addon mk

dimitre commented:
I think the optional subfolders will complicate the regex needed by makefiles. 
The best is having only one folder depth. If any improvement / disambiguating is needed I suggest using a target triplet as a single folder

danoli3 committed:
binutils-gold to depends

danoli3 committed:
apothecary d5cb2e24b30e50e908280b99285746e7beaf0ff4

danoli3 committed:
oF script / linux downloader

danoli3 committed:
Linux Wayland support

danoli3 committed:
glfw

danoli3 committed:
linux pg downloader

danoli3 committed:
linux makefile update arm64 / jetson / rpi

ofTheo commented:
A couple pf questions for this PR to get merged: 
- can we avoid subfolders ( just dealt with issues in emscripten due to this ) 
- can we avoid naming the gcc version ie: ( and just have lib/linux64/myLib.a ) 
- can we  pick a minimum gcc version for libs? ie: gcc10? gcc6 libs have been stable across later platforms. 
- we don't need i386 - 0.12.1 is 64bit only 

I saw you mentioned: 
> Performance Boost for Linux apps 400% for correct GCC
How was that benchmarked?

danoli3 commented:
> A couple of questions for this PR to get merged:
> 
> * can we avoid subfolders ( just dealt with issues in emscripten due to this )
> * can we avoid naming the gcc version ie: ( and just have lib/linux64/myLib.a )
> * can we  pick a minimum gcc version for libs? ie: gcc10? gcc6 libs have been stable across later platforms.
> * we don't need i386 - 0.12.1 is 64bit only
> 
> I saw you mentioned:
> 
> > Performance Boost for Linux apps 400% for correct GCC
> 
> How was that benchmarked?

subfolders are required - linux has two targets aarch64 and 64 by default the developer needs to build for these now. ARM64 is the standard going forward. i.e. all Macs running parallels/vm run arm64 linux 

Not having the subfolders caused the issue as historically we always had subfolders for platforms 
- iOS/armv7/ 
- android/armv7 

For GCC version, if you are thinking for nightly you can set to GCC13 or a minimal one and set download for that.

We can restrict GCC deploys to the majors though:
- Ubuntu 20.04 LTS: Default GCC 10.
- Ubuntu 22.04 LTS: Default GCC 11.
- Ubuntu 24.04 LTS: Default GCC 13 (e.g., gcc-13 13.2.0-23ubuntu4).
- Ubuntu latest: GCC 14 

Bench marks of 400% maybe inflated but based on analysis of some benchmarks of GCC6 vs GCC11 -> 40% increase with a 5-10% optimisation per version. a closer analysis suggests about a 100% increase in performance for x64. 400% for ARM64 linux vs running in EMU mode x64 binaries, however that is actually 800% 

```
Evidence-Based Estimate
Phoronix (GCC 7 vs. GCC 11, April 2021):
Across ~40 benchmarks, GCC 11 was ~20-30% faster than GCC 7 on x86_64 (e.g., FLAC encoding, ray tracing). That’s ~5-7.5% per version over 4 steps.
SPEC CPU 2017 (GCC 6 vs. GCC 10, Intel Data, 2020):
GCC 10 improved SPECint by 25-35% over GCC 6 (4 steps, ~6-8% per version) on x86_64, higher with PGO (40%).
ARM64 Context (GCC 9 to 14):
From GCC 9 to 14 (5 steps), ARM64-specific gains (e.g., SVE, NEON) suggest 40-60% total improvement (8-12% per version), per my prior GCC 14 analysis.
Extrapolating:
GCC 6 to 14 (x86_64): 40-60% total (5-7% per version), based on SPEC and Phoronix trends extended from GCC 11.
GCC 6 to 14 (ARM64): 50-80% total (6-10% per version), due to stronger ARM64 focus in later releases.
Conclusion: GCC 14 vs. GCC 6
Not 10-20% Per Version: A 10-20% gain per release would imply 114-306% total improvement, which exceeds observed data. Real gains are more modest and workload-dependent.
Actual Improvement:
x86_64: GCC 14 is 40-60% faster than GCC 6 (~5-7% per version). Peaks at ~80% in vectorized code.
ARM64: GCC 14 is 50-80% faster than GCC 6 (~6-10% per version), up to 100% in optimized cases (e.g., SVE-enabled loops).
Why Lower Than 10-20%?: Gains diminish as compilers mature; GCC 6 was already decently optimized, and later versions refine rather than revolutionize codegen.
```

However for arm64 with SIMD / Neon this is indeed about a 400% increase with GCC 
```
Emulation Overhead
QEMU Emulation:
Running aarch64 binaries on x86_64 via QEMU (user-mode emulation) incurs significant overhead. Studies (e.g., Linaro, 2020) show QEMU emulation slows execution by 2x to 10x (100-900% slower) compared to native, depending on workload:
Compute-heavy (e.g., math): ~2-3x slower.
I/O or syscall-heavy: ~5-10x slower.
For C++ with GCC 14, assume a 3x slowdown (200%) as a median estimate for mixed workloads.
Native ARM64 Baseline:
A modern ARM64 CPU (e.g., Graviton 3, Apple M1) running GCC 14 aarch64 binaries is already competitive with x86_64. Some benchmarks (e.g., Phoronix, Graviton 3 vs. Zen 4) show ARM64 within 10-20% of x86_64 performance, or better in power-efficient tasks.
GCC 14’s Role
GCC 14 on x86_64 (emulated aarch64) doesn’t benefit from aarch64-specific optimizations, as the binary targets x86_64. The emulation layer (QEMU) interprets aarch64 instructions, negating hardware advantages.
Native ARM64 GCC 14 binaries leverage SVE, NEON, and fused instructions, unavailable in emulation.
Estimated Speed Difference
Native ARM64 (GCC 14): Baseline performance.
x86_64 Emulated (GCC 14): ~3x slower (200% overhead) due to QEMU.
Relative Speed Increase: Native ARM64 is 200-300% faster (3-4x) than emulated x86_64 for typical C++ workloads. For compute-intensive code (e.g., matrix ops), it could be 150-200% faster (2.5-3x), as emulation overhead is lower
```

Basically newer GCC's are closer to Clang performance but we should look at a Clang compiler for Linux for 13 as well after CMake toolchains implemented. Not going to do a clang makefile system changes

danoli3 commented:
Or we can just force apt install of GCC14 for the user ? seems like its worth it and only ship GCC14

danoli3 committed:
Linux rasbian to folder
makefiles fixed
remove i386

danoli3 committed:
make shell

danoli3 committed:
makefiles must use tabs

danoli3 committed:
Linux arm64 fix

danoli3 commented:
![Uploading image.png…]()

Linux arm64 working

danoli3 commented:
Linux 64 also working. Addons probably as broken as the other platforms due to projectGenerator. 
<img width="1165" alt="image" src="https://github.com/user-attachments/assets/0abe3831-f72a-44cd-b93a-a4d5579adeeb" />

Rebaseing was a nightmare, shall merge this very soon.

danoli3 commented:
@dimitre I think the linux in one folder idea works but for the config files. Can do that later this week.

danoli3 commented:
adding setupGCC.sh testing compiling it if no package manager  - 140mb download takes like 15 mins to compile on default GCC if not higher installed. by default it will use package one so instant. more a backup. 
installed after compile GCC is 1.7gb so not worth packaging ourselves 
<img width="541" alt="image" src="https://github.com/user-attachments/assets/0c4cf6ad-7b70-41df-a08d-fa4c9871e291" />

danoli3 committed:
Setup GCC.sh

danoli3 commented:
GCC will automatically be at 10 (lowest) unless installed manually by developer where it detects higher GCC 
<img width="658" alt="image" src="https://github.com/user-attachments/assets/2c94352a-273f-4d7d-8f50-ab24e8324bad" />

For nightly this means it will download GCC10 as we will set it to. no seperate folders, ensuring compatibility

danoli3 commented:
this is the same idea as MSVC 2019 and MSVC 2022 as different GCC's link to different libstdc++ sub versions and compatibility across. It should be tested at later lengths

archlinux is at gcc14

danoli3 committed:
download libs linux minimum gcc 10

danoli3 committed:
nightly arm64 linux

danoli3 committed:
testing nightly

danoli3 committed:
Merge branch 'master' into linux2025

danoli3 committed:
Merge commit '5630d2034d835e81225a225355e97f4144b21179' into linux2025

* commit '5630d2034d835e81225a225355e97f4144b21179':
  Restore target build dir behavior for Xcode. Closes #8321 (#8340)
  of script updates (#8339)
  _OBJC_CLASS_$_AVFoundationVideoPlayer fix (#8337)

# Conflicts:
#	scripts/of.sh

ofTheo commented:
suggest not merging until after 0.12.1 
lot of changes needed in PG front and backend to support the new platform format 

as soon as we get 0.12.1 out we can merge and then make sure new PG tools all work well with this.

tomana commented:
> Just FYI Wayland is good now - working in the nightly builds.

Hi! Running on this branch under native Wayland (GLFW picks `GLFW_PLATFORM_WAYLAND` automatically), the app's `keyPressed`, `mousePressed`, and `mouseMoved` handlers never fire, and the first key press segfaults.

Not sure if _pure_ Wayland is in scope yet, just flagging it.

Tested on Ubuntu 25.10, GNOME 49, GLFW 3.4, GCC 15.2, NVIDIA 580.159.

ofTheo commented:
@tomana - def test the nightly build in the release section of our Github - the Wayland fixes are different there than this PR.  

We should probably close this PR and cherry pick any separate fixes needed.

tomana commented:
@ofTheo As far as I see, nightly is still XWayland. The nice thing about the linux2025 branch is that it allows pure Wayland. I don't see it in the roadmap, but are there plans for supporting pure Wayland?

ofTheo commented:
Heya @tomana - the reason pure Wayland isn't ideal / on the roadmap is that it doesn't allow for setting the window position via the API.  That's a pretty big feature that OF users need. 

If we get a PR with pure Wayland support that covers all the core windowing features like setting window position then we'd definitely consider it.

danoli3 commented:
Okay i agree GCC10 for 12.2 and i'll fix up the make files to auto do whatever can use some special characters and stuff to use all paths we want. 

GCC11 C++20 minimum for Dawn Linux. Angle Linux. So definitely GCC11 for 13.0

danoli3 commented:
I will be closing this when Linux 2026 PR comes in soon
